pkg/etcd/migrate.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/pingcap/log"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	clientV3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+	"go.etcd.io/etcd/server/v3/mvcc"
+	"go.uber.org/zap"
+)
+
+const (
+	// cdcMetaVersion is hard code value indicate the metaVersion of TiCDC
+	cdcMetaVersion       = 1
+	etcdSessionTTL       = 10 * time.Second
+	migrationCampaignKey = "ticdc-migration"
+)
+
+type keys map[string]string
+
+func (k keys) addPair(old, new string) {
+	k[old] = new
+}
+
+type migration struct {
+	oldMetaVersion int
+	newMetaVersion int
+	metaVersionKey string
+	// cdc onwer key in source etcd
+	scrOwnerKey string
+	// destination etcd client
+	desCli           *Client
+	keys             keys
+	srcNoMetaVersion bool
+}
+
+func (m migration) migrate(ctx context.Context) error {
+	// 1.campaign old owner to make sure old keys will not be updates
+	if err := m.campaignOldOwner(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	// 2.campaign owner successfully, begin to migrate data
+	var cmps []clientV3.Cmp
+	var opsThen []clientV3.Op
+	txnEmptyOpsElse := []clientV3.Op{}
+
+	// make sure metaVersion
+	if m.srcNoMetaVersion {
+		cmps = append(cmps, clientV3.Compare(clientV3.CreateRevision(m.metaVersionKey), "=", 0))
+	} else {
+		cmps = append(cmps, clientV3.Compare(clientV3.Value(m.metaVersionKey), "=", fmt.Sprintf("%d", m.oldMetaVersion)))
+	}
+
+	for oldKey, newKey := range m.keys {
+		resp, err := m.desCli.Get(ctx, oldKey, clientV3.WithPrefix())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, v := range resp.Kvs {
+			oldKey := string(v.Key)
+			// make sure newKey doesn't exist
+			cmps = append(cmps, clientV3.Compare(clientV3.CreateRevision(newKey), "=", 0))
+			// make sure oldKey exists
+			cmps = append(cmps, clientV3.Compare(clientV3.ModRevision(oldKey), "!=", 0))
+			// put newKeys on
+			opsThen = append(opsThen, clientV3.OpPut(newKey, string(v.Value)))
+		}
+	}
+
+	// update meataVersion
+	opsThen = append(opsThen, clientV3.OpPut(m.metaVersionKey, fmt.Sprintf("%d", m.newMetaVersion)))
+
+	txnResp, err := m.desCli.Txn(ctx, cmps, opsThen, txnEmptyOpsElse)
+	if err != nil {
+		return cerror.WrapError(cerror.ErrPDEtcdAPIError, err)
+	}
+	if !txnResp.Succeeded {
+		log.Warn("migration compare failed")
+		return cerror.WrapError(cerror.ErrPDEtcdAPIError, err)
+	}
+	log.Info("etcd data migration done")
+	return nil
+}
+
+func (m migration) campaignOldOwner(ctx context.Context) error {
+	sess, err := concurrency.NewSession(m.desCli.Unwrap(),
+		concurrency.WithTTL(int(etcdSessionTTL)))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	election := concurrency.NewElection(sess, m.scrOwnerKey)
+	electionCtx, cancel := context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
+	defer func() {
+		_ = sess.Close()
+	}()
+
+	// TODO(dongmen): do more investigration about how to handle error here
+	if err := election.Campaign(electionCtx, migrationCampaignKey); err != nil {
```

Comment:
should be compaign the cdc owner key?

Reply:
It shouldn't, etcd will campaign on election's prefix key we set as `m.oldOwnerKey` above:
```
election := concurrency.NewElection(sess, m.oldOwnerKey)
```